### PR TITLE
Asyncify filters

### LINKED ISSUE #2573: 'AsyncEth' object has no attribute 'filter'
* Version: 6.0.0b4
* Python: 3.10
* OS: linux
```
aiohttp==3.7.4.post0
async-timeout==3.0.1
attrs==21.2.0
autopep8==1.6.0
base58==2.1.0
bitarray==1.2.2
certifi==2021.10.8
chardet==4.0.0
charset-normalizer==2.0.7
cytoolz==0.11.0
eth-abi==3.0.0
eth-account==0.6.1
eth-hash==0.3.2
eth-keyfile==0.6.0
eth-keys==0.4.0
eth-rlp==0.3.0
eth-typing==3.1.0
eth-utils==2.0.0
hexbytes==0.2.2
idna==3.3
importlib-resources==5.8.0
iniconfig==1.1.1
ipfshttpclient==0.8.0a2
jsonschema==4.7.2
lru-dict==1.1.7
multiaddr==0.0.9
multidict==5.2.0
netaddr==0.8.0
packaging==21.2
parsimonious==0.8.1
pluggy==1.0.0
protobuf==3.19.0
psutil==5.9.0
py==1.10.0
pycodestyle==2.8.0
pycryptodome==3.11.0
pyparsing==2.4.7
pyrsistent==0.18.0
pytest==6.2.5
pytest-asyncio==0.18.3
requests==2.26.0
rlp==2.0.1
six==1.16.0
toml==0.10.2
toolz==0.11.1
typing-extensions==3.10.0.2
urllib3==1.26.7
varint==1.0.2
web3==6.0.0b4
websockets==10.3
yarl==1.7.0
zipp==3.8.1
```

### What was wrong?
It looks like `AsyncEth` module is incapable of log filtration, while `Eth` counterpart is. Here's the [code](https://github.com/mnk-web3/contracts/blob/71865a68bc547e190bd962c8de41348c59cfb1ce/tests/tests/test_check.py#L33), which fails with
```
>       async for event in game_created_builder.deploy(w3).get_all_entries():

tests/test_check.py:33: 
_ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _

self = <web3._utils.events.EventFilterBuilder object at 0x7f9080c8aa70>, w3 = <web3.main.Web3 object at 0x7f9080c8a140>

    def deploy(self, w3: "Web3") -> "LogFilter":
        if not isinstance(w3, web3.Web3):
            raise ValueError(f"Invalid web3 argument: got: {w3!r}")
    
        for arg in AttributeDict.values(self.args):
            arg._immutable = True
        self._immutable = True
    
>       log_filter = cast("LogFilter", w3.eth.filter(self.filter_params))
E       AttributeError: 'AsyncEth' object has no attribute 'filter'

/usr/local/lib/python3.10/site-packages/web3/_utils/events.py:419: AttributeError
```

### PR DESCRIPTION
**Pull request by pacrob**
### What was wrong?

Filters needed asyncing

Related to Issue #1413 
Closes #2573 

### How was it fixed?

Asynced filters

### Todo:
- [x] Add entry to the [release notes](https://github.com/ethereum/web3.py/blob/master/newsfragments/README.md)

#### Cute Animal Picture

image

### CODE REVIEW

File: web3/contract.py [hunk 1]
```diff
@@ -1474,19 +1476,64 @@ def processLog(self, log: HexStr) -> EventData:
...
+        argument_filters: Optional[Dict[str, Any]] = None,
+        fromBlock: Optional[BlockIdentifier] = None,
+        toBlock: Optional[BlockIdentifier] = None,
+        blockHash: Optional[HexBytes] = None,
+    ) -> FilterParams:
+
+        if not self.address:
+            raise TypeError(
+                "This method can be only called on "
+                "an instated contract with an address"
```
Comment (fselmo) [inline web3/contract.py:1491]:
```suggestion
                "an instantiated contract with an address"
```

Unrelated to these changes but I think this is a typo?

File: web3/contract.py [hunk 2]
```diff
@@ -1755,6 +1775,47 @@ async def getLogs(
...
+        *,  # PEP 3102
+        argument_filters: Optional[Dict[str, Any]] = None,
+        fromBlock: Optional[BlockIdentifier] = None,
+        toBlock: BlockIdentifier = "latest",
+        address: Optional[ChecksumAddress] = None,
+        topics: Optional[Sequence[Any]] = None,
+    ) -> AsyncLogFilter:
+        """
+        Create filter object that tracks logs emitted by this contract event.
+        # optional --- update the params descriptions here or remove this line
```
Comment (fselmo) [inline web3/contract.py:1790]: One more copy / paste blip

File: web3/contract.py [hunk 3]
```diff
@@ -1678,6 +1657,47 @@ def getLogs(
...
+        *,  # PEP 3102
+        argument_filters: Optional[Dict[str, Any]] = None,
+        fromBlock: Optional[BlockIdentifier] = None,
+        toBlock: BlockIdentifier = "latest",
+        address: Optional[ChecksumAddress] = None,
+        topics: Optional[Sequence[Any]] = None,
+    ) -> LogFilter:
+        """
+        Create filter object that tracks logs emitted by this contract event.
+        # optional --- update the params descriptions here or remove this line
```
Comment (fselmo) [inline web3/contract.py:1672]: I think this was copy / pasted from my comment in the PR that was merged in here 👀

File: web3/eth.py [hunk 4]
```diff
@@ -359,6 +359,35 @@ def set_contract_factory(
     ) -> None:
         self.defaultContractFactory = contractFactory
 
+    def filter_munger(
```
Comment (fselmo) [inline web3/eth.py:362]: We can make this a static method since it doesn't use the `self` reference.
```python
    @staticmethod
    def filter_munger(
        filter_params: Optional[Union[str, FilterParams]] = None,
        filter_id: Optional[HexStr] = None,
    ) -> Union[List[FilterParams], List[HexStr], List[str]]:
```

File: web3/middleware/filter.py [hunk 5]
```diff
@@ -377,3 +383,281 @@ def middleware(method: RPCEndpoint, params: Any) -> RPCResponse:
...
+
+
+async def async_iter_latest_block_ranges(
+    w3: "Web3",
+    from_block: BlockNumber,
+    to_block: Optional[Union[BlockNumber, LatestBlockParam]] = None,
+) -> AsyncIterable[Tuple[Optional[BlockNumber], Optional[BlockNumber]]]:
+    """Returns an iterator unloading ranges of available blocks
+
+    starting from `fromBlock` to the latest mined block,
```
Comment (fselmo) [inline web3/middleware/filter.py:434]:
```suggestion
    starting from `from_block` to the latest mined block,
```

File: web3/middleware/filter.py [hunk 6]
```diff
@@ -377,3 +383,281 @@ def middleware(method: RPCEndpoint, params: Any) -> RPCResponse:
...
+    until reaching toBlock. e.g.:
```
Comment (fselmo) [inline web3/middleware/filter.py:435]:
```suggestion
    until reaching to_block. e.g.:
```

File: web3/middleware/filter.py [hunk 7]
```diff
@@ -377,3 +383,281 @@ def middleware(method: RPCEndpoint, params: Any) -> RPCResponse:
...
+        elif from_block > latest_block:
+            yield (None, None)
+        else:
+            yield (from_block, latest_block)
+            from_block = BlockNumber(latest_block + 1)
+
+
+async def async_get_logs_multipart(
+    w3: "Web3",
+    startBlock: BlockNumber,
```
Comment (fselmo) [inline web3/middleware/filter.py:459]: We should use `start_block` and `stop_block` here, as well as change these in the synchronous `get_logs_multipart`. This can be in a separate PR if preferred but I think it's a pretty small change that would be simple enough to review if it gets its own commit.

File: web3/contract.py [hunk 8]
```diff
@@ -1678,6 +1657,47 @@ def getLogs(
...
+        """
+        filter_builder = EventFilterBuilder(self._get_event_abi(), self.w3.codec)
+        self._set_up_filter_builder(
+            argument_filters,
+            fromBlock,
```
Comment (fselmo) [inline web3/contract.py:1677]: snake case this if this comment isn't resolved after rebase with snake casing PR.

**Review by fselmo — COMMENTED** [review]
First pass comments... This looks great. Thanks for all that work! We should definitely do another pass at it after rebasing with the snake casing PR, as we discussed.

File: conftest.py [hunk 9]
```diff
@@ -123,4 +98,30 @@ def w3_strict_abi():
...
+
+# --- async --- #
+
+
+def is_async_testrpc_provider(provider):
+    return isinstance(provider, AsyncEthereumTesterProvider)
+
+
+@pytest.fixture()
+def async_skip_if_testrpc():
```
Comment (kclowes) [inline conftest.py:112]: similar question here, I thought we removed `skip_if_testrpc` in one of the previous filter PRs, do we need this still?

**Review by kclowes — COMMENTED** [review]
Thanks for taking this on! It's looking good to me! I had a few small comments, and had similar thoughts around camelCase/snake_case-ing as @fselmo said. And I assume you've checked this a few times already, but it's definitely worth a manual check here to double check filter and async filters work before merging.

File: tests/core/filtering/test_contract_past_event_filtering.py [hunk 10]
```diff
@@ -85,3 +86,102 @@ def test_get_all_entries_returned_block_data(
     assert event_data["transactionIndex"] == txn_receipt["transactionIndex"]
     assert is_same_address(event_data["address"], emitter.address)
     assert event_data["event"] == "LogNoArguments"
+
+
+# --- async --- #
+
+
+@pytest.fixture(scope="module")
+def event_loop():
```
[the remark this answers is not in the document]
Comment (pacrob) [reply tests/core/filtering/test_contract_past_event_filtering.py:95]: I had not tried because it's a shorter fixture. I tried to extract the logic and couldn't get it to work. But I found a single-line solution elsewhere in the codebase and plugged than in.

File: tests/core/filtering/test_contract_topic_filters.py [hunk 11]
```diff
@@ -274,3 +261,208 @@ def test_topic_filters_with_list_arguments(
...
+    return _async_wait_for_block_fixture_logic
+
+
+@pytest_asyncio.fixture(scope="module")
+async def async_wait_for_transaction():
+    return _async_wait_for_transaction_fixture_logic
+
+
+@pytest.fixture(scope="module")
+def event_loop():
```
[the remark this answers is not in the document]
Comment (pacrob) [reply tests/core/filtering/test_contract_topic_filters.py:280]: single line solution found instead of extracting

File: web3/contract.py (hunk 1 above)
Comment (pacrob) [reply web3/contract.py:1491]: yepper!

File: web3/contract.py (hunk 3 above)
Comment (pacrob) [reply web3/contract.py:1672]: got it

File: web3/eth.py (hunk 4 above)
Comment (pacrob) [reply web3/eth.py:362]: scratched

File: web3/middleware/filter.py (hunk 5 above)
Comment (pacrob) [reply web3/middleware/filter.py:434]: :heavy_check_mark:

File: web3/middleware/filter.py (hunk 6 above)
Comment (pacrob) [reply web3/middleware/filter.py:435]: :heavy_check_mark:

File: web3/middleware/filter.py (hunk 7 above)
Comment (pacrob) [reply web3/middleware/filter.py:459]: :heavy_check_mark:

File: conftest.py [hunk 12]
```diff
@@ -46,25 +31,12 @@ def is_testrpc_provider(provider):
     return isinstance(provider, EthereumTesterProvider)
```
[the remark this answers is not in the document]
Comment (pacrob) [reply conftest.py:31]: You might be thinking of:
```python
if not isinstance(w3.provider, EthereumTesterProvider):
        w3.provider = EthereumTesterProvider()
```
that was pulled out of `test_existing_filter_instance`? But I'm not sure about removing `skip_if_testrpc`. It would touch a bunch of new files, so I'll try it as a separate PR off master.

[the remark this answers is not in the document]
Comment (pacrob) [reply conftest.py:31]: Removing `skip_if_testrpc` everywhere else does seem to break things, mostly places where `eth_tester` API endpoints aren't implemented. I'll make a note for further discussion.

File: web3/middleware/filter.py [hunk 13]
```diff
@@ -377,3 +383,281 @@ def middleware(method: RPCEndpoint, params: Any) -> RPCResponse:
...
+            )
+            for item in sublist
+        ]
+
+
+class AsyncRequestBlocks:
+    def __init__(self, w3: "Web3") -> None:
+        self.w3 = w3
+
+    def __await__(self) -> Generator[Any, None, "AsyncRequestBlocks"]:
```
[the remark this answers is not in the document]
Comment (kclowes) [reply web3/middleware/filter.py:579]: aha! right. Thanks for the clarification.

**Comment by kclowes** [comment]
@pacrob I just noticed that we have a bunch more warnings in our tests now as well. Looks like they're related to unclosed event loops in the tests.

File: tests/core/filtering/test_contract_past_event_filtering.py (hunk 10 above)
[the remark this answers is not in the document]
Comment (pacrob) [reply tests/core/filtering/test_contract_past_event_filtering.py:95]: Single-line solution seems to leave event loops unclosed, so I put the original solution back. I got it down to 3 lines though.

File: web3/eth.py [hunk 14]
```diff
@@ -605,6 +634,37 @@ async def get_storage_at(
     ) -> HexBytes:
         return await self._get_storage_at(account, position, block_identifier)
 
+    filter: Method[Callable[..., Awaitable[Any]]] = Method(
```
[the remark this answers is not in the document]
Comment (pacrob) [reply web3/eth.py]: Strangely enough, tightening it here works, but has mypy errors when doing the same for Eth filter.

**Comment by kclowes** [comment]
Sweet! Test failure should be fixed in #2715
